src/anyp/PortCfg.cc

```diff
@@ -55,29 +56,39 @@ AnyP::PortCfg::~PortCfg()
     safe_free(defaultsite);
 }
 
-AnyP::PortCfgPointer
-AnyP::PortCfg::clone() const
+AnyP::PortCfg::PortCfg(const PortCfg &other):
+    next(),
+    s(other.s),
+    transport(other.transport),
+    name(other.name ? xstrdup(other.name) : nullptr),
+    defaultsite(other.defaultsite ? xstrdup(other.defaultsite) : nullptr),
+    flags(other.flags),
+    allow_direct(other.allow_direct),
+    vhost(other.vhost),
+    actAsOrigin(other.actAsOrigin),
+    ignore_cc(other.ignore_cc),
+    connection_auth_disabled(other.connection_auth_disabled),
+    ftp_track_dirs(other.ftp_track_dirs),
+    vport(other.vport),
+    disable_pmtu_discovery(other.disable_pmtu_discovery),
+    workerQueues(other.workerQueues),
+    tcp_keepalive(other.tcp_keepalive),
+    listenConn(),
+    secure(other.secure)
 {
-    AnyP::PortCfgPointer b = new AnyP::PortCfg();
-    b->s = s;
-    if (name)
-        b->name = xstrdup(name);
-    if (defaultsite)
-        b->defaultsite = xstrdup(defaultsite);
-
-    b->transport = transport;
-    b->flags = flags;
-    b->allow_direct = allow_direct;
-    b->vhost = vhost;
-    b->vport = vport;
-    b->connection_auth_disabled = connection_auth_disabled;
-    b->workerQueues = workerQueues;
-    b->ftp_track_dirs = ftp_track_dirs;
-    b->disable_pmtu_discovery = disable_pmtu_discovery;
-    b->tcp_keepalive = tcp_keepalive;
-    b->secure = secure;
+    // to simplify, we only support port copying during parsing
+    assert(!other.next);
+    assert(!other.listenConn);
+}
```

Comment:
> Alex: which PR-related copying method does "less than a full copy"? AFAICT, the copy constructor added in this PR copies everything.

> Amos: Except that next() and listenConn() are not copied. They are default-initialized.

They are effectively copied -- their default initialization values are asserted to be the same as the values of those fields in the object being copied. Thus, the effect of this copy constructor is "everything is copied", as it should be.

If you insist, we can explicitly copy those nil values as well. IMO, the current code is better because its intent is clearer: We cannot meaningfully copy those two fields (and that is OK if and only if there is nothing to copy). The source code comment helps relay that idea.

We can also add assertions that the default-initialized `next` and `listenConn` values (in `this`) are also nil.

I considered also asserting that `next` and `other.next` are the same, but I think that would send the wrong message -- the code should avoid comparing raw port and connection pointers.

Reply:
> > Alex: which PR-related copying method does "less than a full copy"? AFAICT, the copy constructor added in this PR copies everything.
> 
> > Amos: Except that next() and listenConn() are not copied. They are default-initialized.
> 
> They are effectively copied -- their default initialization values are asserted to be the same as the values of those fields in the object being copied. Thus, the effect of this copy constructor is "everything is copied", as it should be.
> 
> If you insist, we can explicitly copy those nil values as well. IMO, the current code is better because its intent is clearer: We cannot meaningfully copy those two fields (and that is OK if and only if there is nothing to copy). The source code comment helps relay that idea.
> 

If you are sure that nothing goes wrong with the case that they are not nil, I will trust you on that and accept.


> We can also add assertions that the default-initialized `next` and `listenConn` values (in `this`) are also nil.
> 

Okay/good.

> I considered also asserting that `next` and `other.next` are the same, but I think that would send the wrong message -- the code should avoid comparing raw port and connection pointers.

Um, it should not matter whether the pointer is raw or not for such a test. Even a safe-pointer comparison is just checking the raw value is identical. So caring enough to impose a "should" is not reasonable IMO.

Reply:
> If you are sure that nothing goes wrong with the case that they are not nil, I will trust you on that and accept.

I am reasonably sure that this code only runs during (re)configuration -- parsePortCfg() -- when the being-copied port is freshly created with those two fields set to nil. The PR code asserts their nil state (and hides port copying from the rest of the code). This is the best we can do for now AFAICT.

Long-term, those fields should not be in the PortCfg class, of course.

> it should not matter whether the pointer is raw or not for such a test

I agree. The word "raw" in my explanation was just misleading noise.